### PR TITLE
First version of the NRF operator

### PR DESCRIPTION
**Pull request by gruyaume**
### Overview

This PR implements the first version of the NRF operator. As of right now the operator supports being deployed and related to a mongo-db-k8s operator. In addition to the charm code, this PR includes tests (unit, integration, linting, static), License, a CI pipeline, and a minimal README.md. Documentation for the charm is minimal as it will mostly be contained in the overarching "charmed 5G" documentation as nobody will simply deploy the NRF on its own.

### CODE REVIEW

**Review by ghislainbourgeois — APPROVED** [review]

File: .github/workflows/codeql-analysis.yml [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+# ******** NOTE ********
+# We have attempted to detect the languages in your repository. Please check
+# the `language` matrix defined below to confirm you have the correct set of
+# supported CodeQL languages.
+#
+name: "CodeQL"
+
+on:
+  push:
+    branches: [ "main" ]
```
Comment (Gmerold) [inline .github/workflows/codeql-analysis.yml:16]: I think we wanna run that check on every push, don't we?

File: src/charm.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+        self.framework.observe(self._database.on.database_created, self._on_database_created)
+        self._service_patcher = KubernetesServicePatch(
+            charm=self,
+            ports=[
+                ServicePort(name="sbi", port=NRF_SBI_PORT),
+            ],
+        )
+
+    def _on_database_created(self, event: DatabaseCreatedEvent) -> None:
+        """Handle database created event."""
```
Comment (Gmerold) [inline src/charm.py:53]: Please add `Args` for consistency.

File: src/charm.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+            database_name=DATABASE_NAME,
+            database_url=database_url,
+            nrf_sbi_port=NRF_SBI_PORT,
+        )
+        self._container.push(path=f"{BASE_CONFIG_PATH}/{CONFIG_FILE_NAME}", source=content)
+        logger.info(f"Pushed {CONFIG_FILE_NAME} config file")
+
+    @property
+    def _config_file_is_written(self) -> bool:
+        """Returns whether the config file was written to the workload container."""
```
Comment (Gmerold) [inline src/charm.py:81]: Please add `Returns` for consistency.

File: src/charm.py [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+        return self._relation_created("database")
+
+    def _relation_created(self, relation_name: str) -> bool:
+        """Returns whether a given Juju relation was crated.
+
+        Args:
+            relation_name (str): Relation name
+
+        Returns:
+            str: Whether the relation was created.
```
Comment (Gmerold) [inline src/charm.py:122]:
```suggestion
            bool: Whether the relation was created.
```

File: src/charm.py [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+        if not self._config_file_is_written:
+            self.unit.status = WaitingStatus("Waiting for config file to be written")
+            return
+        self._container.add_layer("nrf", self._pebble_layer, combine=True)
+        self._container.replan()
+        self.unit.status = ActiveStatus()
+
+    @property
+    def _database_relation_is_created(self) -> bool:
+        """Returns whether database relation is created."""
```
Comment (Gmerold) [inline src/charm.py:112]: Please add `Returns` for consistency.

File: tox.ini [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+setenv =
+  PYTHONPATH = {toxinidir}:{toxinidir}/lib:{[vars]src_path}
+  PYTHONBREAKPOINT=pdb.set_trace
+  PY_COLORS=1
+passenv =
+  PYTHONPATH
+  CHARM_BUILD_DIR
+  MODEL_SETTINGS
+
+[testenv:fmt]
```
Comment (Gmerold) [inline tox.ini:25]: Is anyone actually using it anyway?

**Review by Gmerold — APPROVED** [review]
Other than some nitpicking, LGTM.

File: tox.ini (hunk 6 above)
Comment (sanchezfdezjavier) [reply tox.ini:25]: I am haha. How do you automatically format the code?

File: tox.ini [hunk 7]
```diff
@@ -0,0 +1,82 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
```
Comment (sanchezfdezjavier) [inline tox.ini:3]: I think we should outsource testing dependencies to a file like `test-requirements.txt`

File: tox.ini [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+    -r{toxinidir}/requirements.txt
+commands =
+    coverage run --source={[vars]src_path} -m pytest {[vars]unit_test_path} -v --tb native -s {posargs}
+    coverage report
+
+[testenv:integration]
+description = Run integration tests
+deps =
+    pytest
+    juju<3.1
```
Comment (sanchezfdezjavier) [inline tox.ini:78]: Is there any specific reason for mismatching versions for integration and static?

File: tests/unit/test_charm.py [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+            "password": "password123",
+            "uris": "1.2.3.4:1234,5.6.7.8:1111",
+        }
+        relation_id = self.create_database_relation()
+
+        self.harness.update_relation_data(
+            relation_id=relation_id, app_or_unit="mongodb-k8s", key_values=relation_data
+        )
+
+        patch_push.not_called()
```
Comment (sanchezfdezjavier) [inline tests/unit/test_charm.py:69]: We're not testing that the event is deferred, but the push is not called. Should we change the name of the test?

File: src/charm.py [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+        if not self._database_relation_is_created:
+            self.unit.status = BlockedStatus("Waiting for database relation to be created")
+            return
+        if not self._container.can_connect():
+            self.unit.status = WaitingStatus("Waiting for container to be ready")
+            event.defer()
+            return
+        if not self._config_file_is_written:
+            self.unit.status = WaitingStatus("Waiting for config file to be written")
+            return
```
Comment (sanchezfdezjavier) [inline src/charm.py:105]: Shouldn't we `event.defer()` here?

File: src/charm.py [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+
+        Args:
+            relation_name (str): Relation name
+
+        Returns:
+            str: Whether the relation was created.
+        """
+        if not self.model.get_relation(relation_name):
+            return False
+        return True
```
Comment (sanchezfdezjavier) [inline src/charm.py:126]:
```suggestion
return bool(self.model.get_relation(relation_name))
```

File: tox.ini (hunk 6 above)
Comment (Gmerold) [reply tox.ini:25]: I don't lol :D

File: src/charm.py (hunk 10 above)
Comment (Gmerold) [reply src/charm.py:105]: Not necessarily. Config file is created after DB is created and DB creation will trigger this whole function again anyway.

File: .github/workflows/codeql-analysis.yml (hunk 1 above)
Comment (gruyaume) [reply .github/workflows/codeql-analysis.yml:16]: Well we do run it on pull_request, ain't that good enough? You can [see](https://github.com/canonical/charmed-5g-nrf-operator/actions/runs/4733425656/jobs/8400941502?pr=1) the result of the analysis on this PR for example

Comment (Gmerold) [reply .github/workflows/codeql-analysis.yml:16]: Sure, my point is that if there are some problems, we could potentially detect them faster by running this check on every push.

File: src/charm.py (hunk 10 above)
Comment (gruyaume) [reply src/charm.py:105]: No because on database created (where the config file is created), this same `_configure_pebble_layer` method will be called

Comment (sanchezfdezjavier) [reply src/charm.py:105]: My bad.

File: tests/unit/test_charm.py (hunk 9 above)
Comment (gruyaume) [reply tests/unit/test_charm.py]: Fixed the test name

File: src/charm.py (hunk 11 above)
Comment (gruyaume) [reply src/charm.py]: Done

File: tox.ini (hunk 8 above)
Comment (gruyaume) [reply tox.ini]: Fixed

File: tox.ini (hunk 7 above)
Comment (gruyaume) [reply tox.ini:3]: Done

File: tox.ini (hunk 6 above)
Comment (gruyaume) [reply tox.ini:25]: Resolving this

File: src/charm.py (hunk 5 above)
Comment (gruyaume) [reply src/charm.py]: Done

File: src/charm.py (hunk 4 above)
Comment (gruyaume) [reply src/charm.py]: Done

File: src/charm.py (hunk 3 above)
Comment (gruyaume) [reply src/charm.py]: Done

File: src/charm.py (hunk 2 above)
Comment (gruyaume) [reply src/charm.py]: Done

File: .github/workflows/codeql-analysis.yml (hunk 1 above)
Comment (gruyaume) [reply .github/workflows/codeql-analysis.yml:16]: In all the projects we do it like it is here. If we want to change that we can discuss it but it wouldn't be specific to this project. 

How much faster would it be? I personally haven't felt that running codeql on pull_request made me go slower.

**Review by sanchezfdezjavier — APPROVED** [review]
LGTM

**Pull request merged**